src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotSourceDatasetDependencyDataGcpStep.java

```diff
@@ -0,0 +1,62 @@
+package bio.terra.service.snapshot.flight.delete;
+
+import bio.terra.common.FlightUtils;
+import bio.terra.common.exception.PdaoException;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.dataset.exception.DatasetNotFoundException;
+import bio.terra.service.dataset.flight.DatasetWorkingMapKeys;
+import bio.terra.service.dataset.flight.ingest.OptionalStep;
+import bio.terra.service.filedata.google.firestore.FireStoreDependencyDao;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import com.google.cloud.bigquery.BigQueryException;
+import java.util.UUID;
+import java.util.function.Predicate;
+
+public class DeleteSnapshotSourceDatasetDependencyDataGcpStep extends OptionalStep {
+  private FireStoreDependencyDao dependencyDao;
+  private UUID snapshotId;
+  private DatasetService datasetService;
+
+  public DeleteSnapshotSourceDatasetDependencyDataGcpStep(
+      FireStoreDependencyDao dependencyDao,
+      UUID snapshotId,
+      DatasetService datasetService,
+      Predicate<FlightContext> doCondition) {
+    super(doCondition);
+    this.dependencyDao = dependencyDao;
+    this.snapshotId = snapshotId;
+    this.datasetService = datasetService;
+  }
+
+  @Override
+  public StepResult doOptionalStep(FlightContext context) throws InterruptedException {
+    FlightMap map = context.getWorkingMap();
+    UUID datasetId = map.get(DatasetWorkingMapKeys.DATASET_ID, UUID.class);
+    try {
+      Dataset dataset = datasetService.retrieve(datasetId);
+      dependencyDao.deleteSnapshotFileDependencies(dataset, snapshotId.toString());
+
+    } catch (BigQueryException ex) {
+      if (FlightUtils.isBigQueryIamPropagationError(ex)) {
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+      }
+      throw new PdaoException("Caught BQ exception while deleting snapshot", ex);
+    } catch (DatasetNotFoundException nfe) {
+      // If we do not find the snapshot or dataset, we assume things are already clean
```

Comment:
In general we should be using exceptions for exceptional situations. If it's normal for a dataset to be not found in this situation, it would be better if we didn't have to catch an exception to determine this. Something for a future change, perhaps.

Reply:
Given that we're only running this step if it passes this check: SnapshotDeletePredicates::performGCPDatasetDependencyStep, we should never hit this exception. I'll remove the catch.